src/creator.rs

```diff
@@ -98,53 +98,265 @@ impl<H: Hasher> Creator<H> {
         }
     }
 
-    fn check_ready(&self) -> bool {
-        if self.current_round == 0 {
-            return true;
-        }
-        if self.current_round > self.max_round {
-            warn!(target: "AlephBFT-creator", "{:?} Maximum round reached. Not creating another unit.", self.node_ix);
-            return false;
+    async fn wait_until_ready(&mut self, round: Round) {
+        let prev_round_index = match round.checked_sub(1) {
+            Some(prev_round) => prev_round as usize,
+            None => {
+                Delay::new((self.create_lag)(round.into())).await;
+                return;
+            }
+        };
+
+        let mut delay = Delay::new((self.create_lag)(round.into())).fuse();
+        loop {
+            // We need to require a number higher by one then currently highest round
+            // (by 2 then length) to prevent attack when a malcious node is creating
+            // units without any delay to achive max_round as soon as possible
+            if ((round + 2) as usize) < self.n_candidates_by_round.len() {
+                // Since we get unit from round r, we have enough units from previous
+                // rounds to skip delay, because we are already behind
+                return;
```

Comment:
This is incorrect now. Our unit from previous round must also be available to return here.

Reply:
To clarify: the fact that we have created a unit is not the same thing as having it available as a parent in the Creator -- the created unit must first go through several components to "come back" to the Creator.

Reply:
Done